### DISCUSSION TITLE
Device IO function redesign

### DISCUSSION THREAD

**Discussion by jyao1** [DMTF/libspdm#702] [Ideas]
Here is the discussion for https://github.com/DMTF/libspdm/issues/351

[Device IO Function Redesign.pdf](https://github.com/DMTF/libspdm/files/8199075/Device.IO.Function.Redesign.pdf)

Update the foil according to the meeting discussion.

**Comment by steven-bellock** [comment]
@jyao1 should we discuss this at tomorrow meeting?

**Comment by jyao1** [comment]
Yes, let's discuss it

**Comment by jyao1** [comment]
@richkong88, @ekohandel, please share your idea too

**Reply by ekohandel** [reply]
Sorry I wasn't able to attend the meeting. I didn't see this update until it was too late. I have looked through the proposal and it looks really good. Thank you @jyao1 for putting that together.

This is not strictly needed for this proposal but it is in the spirit of buffer size reduction. In a scenario where a single compilation of the library is used for transports with different receive and transmit buffer sizes, one has to choose the buffer sizes for the scratch and last request buffers to match the largest of the multiple transports. It would be ideal if these buffers are set at runtime with a size that is also provided at runtime so one could tailor each context to the transport it is paired with.

**Comment by jyao1** [comment]
@richkong88, @ekohandel, we just updated the code to include this feature. Would you please take a look?

**Reply by richkong88** [reply]
Looks good. Only one small comment. It looks like a crash/undefined behavior is possible on allocation failure, as only the message_size is checked via assert and the return_status is ignored. Perhaps we should be checking the return_status with an if statement as well?

**Reply by jyao1** [reply]
yes, good suggestion. I will add error check.
